### PR TITLE
Simplify build yml

### PR DESCRIPTION
**Pull request by scbedd**
## Setting some context

This change is best explained in pictures. A standard build looks like this:

image

With the job dependencies as shown

```
Build
  -> Build_Extended
    -> Analyze
  -> TestN
```

When an `extension` package is present, this changes. (indicated by `AdvancedBuild` boolean from `ci.yml`)

```
Build_Linux
Build_Windows
Build_Mac
  -> Coalesce Artifacts
    -> Build_Extended
      -> Analyze
    -> TestN
```

As you can see. The build is bigger when we _know_ we have an extension package present. But what about when we **don't** know at template compile time if we have an extension package? That's the world we're heading for with the `pipelinev3` efforts.

This PR is setting the stage to allow a PR build which expands and contracts to **handle** when `azure-storage-extensions` is detected to be part of the necessary build set.

## What this PR does

- [x] Removes compile-time difference between `advanced` and `non-advanced` builds
- [x] Allows run-time detection of `storage` service directory to produce non-linux artifacts when `azure-storage-extension` is detected
- [x] Uses `ENABLE_EXTENSION_BUILD` to skip the build tasks on `windows` and `mac` when `azure-storage-extension` (or any extension package later) is NOT detected

And the job dependencies will look like:

```
Build_Linux
Build_Windows (steps skipped if no extension package)
Build_Mac (steps skipped if no extension package)
  -> Build_Extended
    -> Analyze
  -> TestN
```

### CODE REVIEW

File: eng/pipelines/templates/steps/build.yml [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+steps:
+    - template: /eng/pipelines/templates/steps/targeting-string-resolve.yml
+      parameters:
+        BuildTargetingString: ${{ parameters.BuildTargetingString }}
+
+    - template: /eng/pipelines/templates/steps/resolve-build-platforms.yml
+      parameters:
+        ServiceDirectory: ${{ parameters.ServiceDirectory }}
+
+    - template: ../steps/build-package-artifacts.yml
```
Comment (weshaggard) [inline eng/pipelines/templates/steps/build.yml:30]:
```suggestion
    - template: /eng/pipelines/templates/steps/build-package-artifacts.yml
```

File: eng/pipelines/templates/steps/download-package-artifacts.yml [hunk 2]
```diff
@@ -0,0 +1,15 @@
+steps:
+    - task: DownloadPipelineArtifact@2
+      inputs:
+        artifactName: 'packages_windows'
```
[the remark this answers is not in the document]
Comment (weshaggard) [reply eng/pipelines/templates/steps/download-package-artifacts.yml:4]: I read your description in more detail now and I guess we do build platform specific for azure-storage-extension. That makes things more complicated :(

File: eng/pipelines/templates/steps/resolve-build-platforms.yml [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+    default: ''
+
+steps:
+    # when we merge pipeline v3, this check will change to examining the targeting string $(TargetingString)
+    # as the generate-pr-diff call + resolution will be present in targeting-string-resolve.yml.
+    # until then, we simply check to see if we're targeting storage service directory
+    - pwsh: |
+        if ("${{ parameters.ServiceDirectory }}" -eq "storage") {
+          Write-Host "Targeting storage, enabling extension build."
+          Write-Host "##vso[task.setvariable variable=ENABLE_EXTENSION_BUILD]true"
```
Comment (weshaggard) [inline eng/pipelines/templates/steps/resolve-build-platforms.yml:13]: Where is this variable consumed?

Comment (scbedd) [reply eng/pipelines/templates/steps/resolve-build-platforms.yml:13]: It's coming! I have a todo checkbox not checked yet. Going to knock it out now. 👍

Comment (scbedd) [reply eng/pipelines/templates/steps/resolve-build-platforms.yml:13]: @weshaggard g2g now.!

This is core, note that the windows/mac are shorter:
image

And this is storage:

**Review by mccoyp — APPROVED** [review]

**Pull request merged**